src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java

```diff
@@ -414,69 +418,111 @@ private void restoreSchemaMetadata(Connection conn, Long schemaID) throws SQLExc
 	}
 
-	private void restoreFullSchema(Connection conn, Long schemaID) throws SQLException, InvalidSchemaError {
-		PreparedStatement p = conn.prepareStatement("SELECT * from `databases` where schema_id = ? ORDER by id");
-		p.setLong(1, this.schemaID);
-
-		ResultSet dbRS = p.executeQuery();
 
-		while (dbRS.next()) {
-			this.schema.addDatabase(restoreDatabase(conn, dbRS.getInt("id"), dbRS.getString("name"), dbRS.getString("charset")));
-		}
-
-	}
-
-	private Database restoreDatabase(Connection conn, int id, String name, String charset) throws SQLException {
-		Statement s = conn.createStatement();
-		Database d = new Database(name, charset);
-
-		ResultSet tRS = s.executeQuery("SELECT * from `tables` where database_id = " + id + " ORDER by id");
-
-		while (tRS.next()) {
-			String tName = tRS.getString("name");
-			String tCharset = tRS.getString("charset");
-			String tPKs = tRS.getString("pk");
+	private void restoreFullSchema(Connection conn, Long schemaID) throws SQLException, InvalidSchemaError {
+		PreparedStatement p = conn.prepareStatement(
+				"SELECT " +
+						"d.id AS dbId," +
+						"d.name AS dbName," +
+						"d.charset AS dbCharset," +
+						"t.name AS tableName," +
+						"t.charset AS tableCharset," +
+						"t.pk AS tablePk," +
+						"t.id AS tableId," +
+						"c.column_length AS columnLength," +
+						"c.enum_values AS columnEnumValues," +
+						"c.name AS columnName," +
+						"c.charset AS columnCharset," +
+						"c.coltype AS columnColtype," +
+						"c.is_signed AS columnIsSigned " +
+						"FROM `databases` d " +
+						"LEFT JOIN tables t ON d.id = t.database_id " +
+						"LEFT JOIN columns c ON c.table_id=t.id " +
+						"WHERE d.schema_id = ? " +
+						"ORDER BY d.id, t.id, c.id"
+		);
 
-			int tID = tRS.getInt("id");
+		p.setLong(1, this.schemaID);
+		ResultSet rs = p.executeQuery();
 
-			restoreTable(conn, d, tName, tID, tCharset, tPKs);
-		}
-		return d;
-	}
+		Database currentDatabase = null;
+		Table currentTable = null;
+		int columnIndex = 0;
+		long dbStart = 0;
+
+		while (rs.next()) {
+			// Database
+			String dbName = rs.getString("dbName");
+			String dbCharset = rs.getString("dbCharset");
+
+			// Table
+			String tName = rs.getString("tableName");
+			String tCharset = rs.getString("tableCharset");
+			String tPKs = rs.getString("tablePk");
+
+			// Column
+			String columnName = rs.getString("columnName");
+			int columnLengthInt = rs.getInt("columnLength");
+			String columnEnumValues = rs.getString("columnEnumValues");
+			String columnCharset = rs.getString("columnCharset");
+			String columnType = rs.getString("columnColtype");
+			int columnIsSigned = rs.getInt("columnIsSigned");
+
+			if (currentDatabase == null || !currentDatabase.getName().equals(dbName)) {
+				if (currentDatabase != null) {
+					long end = System.currentTimeMillis();
+					LOGGER.debug("Processed database " + dbName + " after " + (end - dbStart) + "ms");
```

Comment:
thinking the timing info is kinda not-important even for debugging purposes?  Maybe kill it?